### ISSUE TITLE
Encoding issue with emoji

### ISSUE THREAD

**Issue by Jay2k1** [buxxi/znc-mysql#4]
I know I know, emojis in IRC... but they happen.

I upgraded from Ubuntu 14.04 to 16.04 and now whenever someone posts something containing an emoji, *sql sends this:
```Could not save SAY to database caused by: <class 'pymysql.err.InternalError'> (1366, "Incorrect string value: '\\xF0\\x9F\\x98\\x82' for column 'message' at row 1")```

I did some research and [found out that apparently the MySQL utf8 encoding can't store 4-byte strings](https://stackoverflow.com/a/24559308) and hence throws this error. Using utf8mb4 is said to be the solution, working from MySQL 5.5 upwards.

I was wondering why I didn't get this bug with the old OS and actually grepped for an emoji in my text log files, and then looked up a message containing an emoji in the SQL DB. Sure enough, the message was there, just without an emoji, as if it had been removed.

I think this is responsible for that: https://github.com/buxxi/znc-mysql/blob/master/sql.py#L112

I do wonder why this used to work before. Maybe this line replaced characters python considered invalid, and a newer python version (due to OS upgrade) is now compatible with emoji and passes them on to MySQL which then throws the error.

**Comment by buxxi** [comment]
More encoding problems, how fun! 😂

My server is also running Ubuntu (16.04), but I have started using MariaDB instead of MySQL.
The table has the type: utf8_general_ci

I typed the emoji that I pasted above and the resulting message is an empty text for me.

**Comment by Jay2k1** [comment]
What's your version?
```
Jay2k1@jay2k1 ~ % python --version
Python 2.7.12
```
And it might indeed be a difference between MySQL and MariaDB...

**Comment by buxxi** [comment]
Well, ZNC requires python3 for the python modules.

>[~] python3 --version
>Python 3.5.2

And searching the apt I get this

>[~] apt-cache search pymysql
>python-pymysql - Pure-Python MySQL driver - Python 2.x
>python3-pymysql - Pure-Python MySQL Driver - Python 3.x`

So maybe you should have the second one instead?

**Comment by Jay2k1** [comment]
My python3 version is 3.5.2 too, unsurprisingly. And I have both of these pymysql packages installed.

Then again, even while omitting emojis (and other four-byte unicode characters) would restore the old state, wouldn't it be preferable to actually fix the issue and store those too? That would probably really require a column encoding (and collation) of utf8mb4. And since you have to specify the encoding in your module, maybe it needs to be changed there too.

And because other users might stumble upon this, I think there should be a note of warning (or a fix as in automatic table creation in the module, for example), what do you think?

**Comment by buxxi** [comment]
Of course it would be best to fix it, but I'm not sure how...

And you really think there are more users? I'm surprised there's someone more except me using it :P

**Comment by Jay2k1** [comment]
Heh, I don't know how many users there are. Don't underestimate the usefulness of your plugin though :)

I "fixed" it by converting the message column to utf8mb4 and changing the encoding in line 112 from ```UTF8``` to ```utf8mb4```. Now the error doesn't appear anymore, but a test emoji (this: 😂) ended up as ```?``` in the database. Not optimal, but better than error queries by *sql...

**Comment by buxxi** [comment]
I committed those changes (and made the modules show up in settings).
And for me the emoji is in the actual database after that (had trouble first with the sql-editor I used to see it, only showed up as '?' as you describe) but using the CLI directly on the server with utf-8 locale set I can see that it's there. 

Also tried to remove the encode-decode-code to see if it worked without it and it does, so I may remove that if I don't run in to any charset errors in the near future.
